keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,790 @@
+
+Another area of interest is to avoid autoscaling if a pod is terminating.  In some cases a pod will be recreated where there is capacity and if there is no capacity autoscaling could be triggered.  This is not ideal if a pod is terminating and a user wants to wait until the pod is fully terminated.
+
+### Goals
+
+- Deployment controller should allow for flexibility in waiting for pods to be fully terminated before
+  creating new ones
+- Deployment/ReplicaSet will have a new status field where we include the number of terminating replicas.
+
+### Non-Goals
+
+- DaemonSets and StatefulSets are not included in this proposal
+  - They were designed to enforce uniqueness from the start so we will not include them in this design.
+
+## Proposal
+
+ReplicaSet controller get a list of active pods.  Active pods usually mean pods that have not been registered for deletion.  In this KEP, we will consider terminating pods to be separate from active and failed.  This means that for cases where we track the number of pods, we should include a field that states the number of terminating pods.  
+
+We will propose a new API fields in Deployments/ReplicaSets in this KEP.  
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running includes terminating pods.  In scare compute environments, users may only have a limited amount of nodes and they do not want to try and schedule pods to a new resource.
+Counting terminating pods as active allows for the scheduling of pods to wait until pods are terminated.
+
+See [Option for acknowledging terminating Pods in Deployment rolling update](https://github.com/kubernetes/kubernetes/issues/107920)
+for more examples.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Open Questions on Deployment Controller
+
+The Deployment API is open for discussion.  We put the field in Deployment/ReplicaSet because it is related to RolloutStrategy.
+It is not clear if `recreate` and/or `rollingupdate` need this API for both rollout options.
+
+Another open question is if we want to include Deployments in the initial release of this feature.  There is some discussion about releasing the Job API first and then follow up with Deployment.  
+
+We decided to define the APIs in this KEP as they can utilize the same implementation.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API Name Choices
+
+- TerminatingAsActive
+- ActiveUntilTerminal
+- DelayPodRecreationUntilTerminal
+- ?
```

Comment:
choices for the field and values: https://github.com/kubernetes/enhancements/pull/3940#discussion_r1195616467